### DISCUSSION TITLE
MappedClassProtocol equivalent for type vars

### DISCUSSION THREAD

**Discussion by DavideCanton** [sqlalchemy/sqlalchemy#9710] [Usage Questions]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Disclaimer: maybe this is not strictly a typing issue, but more a suggestion, not sure this is the right place to report it.

It seems that `MappedClassProtocol` cannot be used as bound for `TypeVar`s, since it requires a `__call__` method, which orm instances typically don't define.

In the reproduce example, when `T` is instantiated with `MyModel`, which is a mapped class, mypy errors because instances of `MyModel` do not define a `__call__` method.

On the other hand, instantiating the typevar with `type[MyModel]` works, but then I can't use `MyModel` inside the generic class, since `T` would be substituted with `type[MyModel]`.

I believe the protocol was defined this way because it was thought for model classes, not for instances, but maybe it may make sense having another protocol specific for orm instances?

### To Reproduce

```python
from abc import ABCMeta, abstractmethod
from typing import Generic, TypeVar
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedClassProtocol, registry

Registry = registry()


class Base(DeclarativeBase):
    registry = Registry


class MyModel(Base):
    __tablename__ = "my_table"

    id: Mapped[int]


T = TypeVar("T", bound=MappedClassProtocol)


class Handler(Generic[T], metaclass=ABCMeta):
    @abstractmethod
    def handle(self, model: T) -> None:
        pass


class MyModelHandler(Handler[MyModel]):
    def handle(self, model: MyModel) -> None:
        pass
```

### Error

```
foo.py:27:22: error: Type argument "MyModel" of "Handler" must be a subtype of "MappedClassProtocol[Any]"  [type-var]
``` 

### Versions

- OS: Windows 10
- Python: 3.10
- SQLAlchemy: 2.0.4
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy 1.1.1

**Comment by zzzeek** [comment]
`MappedClassProtocol` is a protocol that itself defines a "type"; it's not a real "Type[]" due to the way we use it, but it would never be the target of a `TypeVar` since it's already intended to act as a "Type", that's why it has `__call__()`.   pep-484 IIUC doesn't really have protocols for "classes with class-level attributes", or i was possibly confused, but that's what MappedClassProtocol is right now.

If you are creating a handler for a hierarchy of classes, the base class of all that here is your `Base` itself:

```
T = TypeVar("T", bound=Base)
```

if you're trying to do something with any arbitrary SQLAlchemy mapped class from a third party, SQLAlchemy can apply mappings to anything that extends from `object`.    Since protocols use duck typing anyway there's no reason you couldn't make a new protocol that meets your needs like this:

```py
_O = TypeVar("_O")

class MappedProtocol(Protocol[_O]):
    """A protocol representing a SQLAlchemy mapped instance.

    """

    __mapper__: Mapper[_O]
    __table__: FromClause

T = TypeVar("T", bound=MappedProtocol)


```

**Reply by DavideCanton** [reply]
Yes, that was my understanding too.

Using `Base` directly is out of question unfortunately since I have to combine it with other protocols, so I already went for something like this:

```py
class Model(Protocol):
    __table__: FromClause

class ModelWithField(Model, Protocol):
    field: Mapped[int]


T = TypeVar("T", bound=Model)
F = TypeVar("F", bound=ModelWithField)


class Handler(Generic[T], metaclass=ABCMeta):
    @abstractmethod
    def handle(self, model: T) -> None:
        pass


class BaseHandler(Handler[T]):
    def handle(self, model: T) -> None:
        pass


class FieldHandler(BaseHandler[F]):
    def handle(self, model: F) -> None:
        pass

```

Python doesn't allow mixing protocols and non-protocols when subclassing, and typevars do not support intersection of types, so this is the only way to support my use case.

I was wondering if there was some other utility protocol inside sqlalchemy to support my case, but I didn't find anything related.

**Reply by zzzeek** [reply]
all of our typing stuff was written just so that our own APIs are typed and that end-user types can be exposed as much as possible, but in the realm of types that SQLAlchemy exports for end-user use, we have nearly none right now, though @caselit is working on adding some.

**Reply by DavideCanton** [reply]
ok, guess I'll stick with my custom protocol until something better comes out :)

Thanks for your time!
